### ISSUE TITLE
Global variables in yaml files

### ISSUE THREAD

**Issue by armab** [StackStorm/st2#1651]
There are already some [environment variables available to the actions](https://github.com/StackStorm/st2/blob/master/docs/source/actions.rst#common-environment-variables-available-to-the-actions).
Most probably there will be more and more useful of these during time.

The proposal is to make those variables available in yaml files too, bringing more flexibility.
```yaml
---
name: example_action
runner_type: local-shell-cmd
entry_point: ""
enabled: true
parameters:
  cwd:
    default: "/opt/stackstorm/virtualenvs/{{ ST2_ACTION_PACK_NAME }}/bin"
    immutable: true
  cmd:
    ...
```

Btw, this approach is pretty obvious thing in [Ansible](https://docs.ansible.com/playbooks_variables.html#information-discovered-from-systems-facts), which operates with yaml files too.

**Comment by Kami** [comment]
That's related to allowing users to reference values from a datastore in the parameter default values - we decided to keep it explicit, simple and easier to debug by not allowing it.

If user wants to reference a datastore value, they need to add a task in a workflow which retrieves this value - this way it's explicit and easier to see what is going on and where value is coming from.

I hold the same opinion for using environment variable values in the parameter default values - I prefer to keep metadata simple and static. There is obviously a lot of dynamic stuff going on in the actions, workflows and rules, but I think it's better if we try to keep metadata files as "static" as possible.

**Comment by armab** [comment]
I see, thanks for explanation.

From my understanding allowing to make things in less steps without writing dirty hacks could be good.
I hope it's just matter of time :calendar:  when users would ask more and more flexibility ;)

**Comment by armab** [comment]
Here is another example I've experienced:

When pack is renamed, since it's name is hardcoded in yaml files - everything inside the pack should be renamed too.

Assume I have `st2-chatops-pack` with actions and aliases in it.
Action: `st2-chatops-pack/actions/cowsay.yaml`
Action alias: `st2-chatops-pack/aliases/cowsay.yaml`:
```yaml
---
name: chatops.cowsay
# could be:
# action_ref: {{ ST2_ACTION_PACK_NAME }}.cowsay
action_ref: st2-chatops-pack.cowsay
description: Draws a cow that says what you want
formats:
  - "cowsay {{message}}"
```

If I rename the pack I should dig into every yaml file and rename old pack name with new name: `action_ref: new-chatops-pack.cowsay`. Another example situation when I share GitHub repo which someone forked and renamed it to whatever they want.
Renaming could be annoying.

**Comment by Kami** [comment]
@armab I see what you mean, but I don't really think it's a big issue and you can't fully avoid it even if we would support "magic" variables.

In alias definitions you can reference an action from a different pack which means you need to provide a unique identifier (primary key) for an action - in our case that's `<pack name>.<action name>`.

Even if we provided a "magic" variable or allowed user to avoid pack name if they want to reference an action inside the same pack (not a fan of that approach since it's not explicit), the reference would still need to be updated in other packs which reference that action.

So in short - I don't think there is much we can do and the same problem exists everywhere where you change the value of a primary key (that's also usually why the values of primary keys shouldn't change).

**Comment by armab** [comment]
Yeah, that's true.
But not so frequent case. We currently have aliases for [`hubot`](https://github.com/StackStorm/st2contrib/tree/master/packs/hubot/aliases) and [`packs`](https://github.com/StackStorm/st2/tree/master/contrib/packs/aliases), which pointing to the same pack actions.

With this kind of magic the pack could be self-sufficient and just renaming perhaps may provide less difficulties at least inside the pack scope. But for now it's rare example too.

---
Such examples could be signs of possible improvements. I'll post here if would experience more cases like that - may be a good material to consider in future.

**Comment by vincent-legoll** [comment]
Is there a way to reference pack directory from pack files themselves ?
Because using ".." does not work, at least in my tries.

I searched a lot for this and found nothing.

Something that would contain `/opt/stackstorm/pack/{{ ST2_ACTION_PACK_NAME }}`
Or just `/opt/stackstorm/pack` would work too.

**Comment by bigmstone** [comment]
@vincent-legoll You should be able to access pack name via `{{action_context.pack}}`

**Comment by vincent-legoll** [comment]
@bigmstone : I think you misunderstood my request (maybe I was not clear enough), what I want to know is the other part. I already have my pack name, what I need to get is the packs directory, i.e.: `/opt/stackstorm/pack`, so that I can add my pack name to it.

**Comment by cognifloyd** [comment]
Yes. it still makes sense to make a set of variables (like the environment variables) available in action metadata. If all of them are accessible via the `action_context` that's would work.

**Comment by cognifloyd** [comment]
Still an issue.

**Comment by relishcolouredhat** [comment]
I also agree this set of variables would be useful.

Specifically something that can be used to reference the pack working directory, and ideally which would work regardless of if the pack is installed into `/packs` or is being worked on via `/packs.dev`

It seems I could set this in a way that would work for one or the other by using 
` /opt/stackstorm/pack/{{ ST2_ACTION_PACK_NAME }} ` or
` /opt/stackstorm/packs.dev/{{ ST2_ACTION_PACK_NAME }} `

However it seems unclear how I might determine which of the two the pack may be installed into. 

This particular pack is for internal use only, but I'd like to try and leverage as much of the 'pack management infrastructure' as possible.
